vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java

```diff
@@ -671,30 +746,56 @@ public void setValue(LocalDate value) {
         boolean isValueRemainedEmpty = isOldValueEmpty && isNewValueEmpty;
         boolean isInputValuePresent = isInputValuePresent();
 
-        // When the value is cleared programmatically, reset hasInputValue
-        // so that the following validation doesn't treat this as bad input.
+        // When the value is cleared programmatically, there is no change event
+        // that would synchronize _inputElementValue, so we reset it ourselves
+        // to prevent the following validation from treating this as bad input.
         if (isNewValueEmpty) {
-            getElement().setProperty("_hasInputValue", false);
+            setInputElementValue("");
         }
 
         super.setValue(value);
 
-        // Clear the input element from possible bad input.
+        // Revalidate if setValue(null) didn't result in a value change but
+        // cleared bad input
         if (isValueRemainedEmpty && isInputValuePresent) {
-            // The check for value presence guarantees that a non-empty value
-            // won't get cleared when setValue(null) and setValue(...) are
-            // subsequently called within one round-trip.
-            // Flow only sends the final component value to the client
-            // when you update the value multiple times during a round-trip
-            // and the final value is sent in place of the first one, so
-            // `executeJs` can end up invoked after a non-empty value is set.
-            getElement()
-                    .executeJs("if (!this.value) this._inputElementValue = ''");
             validate();
             fireValidationStatusChangeEvent();
         }
     }
 
+    @Override
+    protected void setModelValue(LocalDate newModelValue, boolean fromClient) {
+        // Ignore setModelValue calls triggered by setPresentationValue
+        // when the fallback parser applies a parsed value (see below).
+        // This ensures that the ValueChangeEvent fires from the original
+        // setModelValue call with `fromClient` value: `true`.
+        if (isFallbackParserRunning) {
+            return;
+        }
+
+        boolean isInputUnparsable = fromClient && newModelValue == null
+                && isInputValuePresent();
+        if (fallbackParser != null && isInputUnparsable) {
+            isFallbackParserRunning = true;
+
+            Result<LocalDate> result = fallbackParser
+                    .apply(getInputElementValue());
+            Objects.requireNonNull(result, "Result cannot be null");
+
+            if (result.isError()) {
+                fallbackParserErrorMessage = result.getMessage().orElse(null);
+            } else {
+                fallbackParserErrorMessage = null;
+                newModelValue = result.getOrThrow(IllegalStateException::new);
+                setPresentationValue(newModelValue);
+            }
+
+            isFallbackParserRunning = false;
```

Comment:
Good idea, thanks! Updated. We could even go further, intercept exceptions, and convert them to logger errors, but I'm not sure what to show to the user, bad input error message from i18n?

Reply:
Logging the exception would be a good idea IMO. Could just be a message like "Fallback parser threw an exception" together with the exception that was caught.

For the validation error, maybe just use the I18N bad input message? That seems the most reasonable instead of trying to come up with something new.